lib/node_modules/@stdlib/repl/lib/multiline_handler.js

```diff
@@ -0,0 +1,561 @@
+		this._multiline.active = false;
+	}
+});
+
+/**
+* Resets input buffers.
+*
+* @private
+* @name _resetInput
+* @memberof MultilineHandler.prototype
+* @type {Function}
+* @returns {void}
+*/
+setNonEnumerableReadOnly( MultilineHandler.prototype, '_resetInput', function resetInput() {
+	this._cmd.length = 0;
+	this._lineIndex = 0;
+	this._lines.length = 0;
+});
+
+/**
+* Updates current input line in buffer.
+*
+* @name updateLine
+* @memberof MultilineHandler.prototype
+* @type {Function}
+* @param {string} line - updated line
+* @returns {void}
+*/
+setNonEnumerableReadOnly( MultilineHandler.prototype, 'updateLine', function updateLine( line ) {
+	this._lines[ this._lineIndex ] = line;
+});
+
+/**
+* Processes input line data.
+*
+* @name processLine
+* @memberof MultilineHandler.prototype
+* @type {Function}
+* @param {string} line - line data
+* @returns {void}
+*/
+setNonEnumerableReadOnly( MultilineHandler.prototype, 'processLine', function processLine( line ) {
+	var code;
+	var node;
+	var ast;
+	var cmd;
+	var tmp;
+	var dy;
+
+	debug( 'Line: %s', line );
+
+	// Check for manual newline input...
+	if ( this._multiline.active && this._multiline.mode === 'manual' ) {
+		debug( 'Detected newline input via modifier-key. Waiting for additional lines...' );
+
+		// Update current line:
+		this._cmd[ this._lineIndex ] = line;
+
+		// Insert a newline:
+		this._lineIndex += 1;
+		this._cmd.splice( this._lineIndex, 0, this._remainingLine );
+		this._lines.splice( this._lineIndex, 0, this._remainingLine );
+
+		// Insert buffer input after cursor from previous line:
+		displayPrompt( this._repl, false );
+		this._ostream.write( this._remainingLine );
+		readline.cursorTo( this._ostream, 0 );
+		this._rli.line = this._remainingLine;
+
+		// Clear buffer:
+		this._remainingLine = '';
+		this._multiline.mode = 'incomplete_expression'; // reset flag
+		return;
+	}
+	this._multiline.active = false; // false until proven otherwise
+	this._cmd[ this._lineIndex ] = line;
+	cmd = this._cmd.join( '\n' );
+	if ( RE_WHITESPACE.test( cmd ) ) {
+		displayPrompt( this._repl, false );
+		return;
+	}
+	if ( RE_SINGLE_LINE_COMMENT.test( cmd ) || RE_MULTI_LINE_COMMENT.test( cmd ) ) { // eslint-disable-line max-len
+		debug( 'Detected single-line comment.' );
+		tmp = cmd;
+	} else {
+		// Check if the command has valid syntax...
+		debug( 'Processing command...' );
+		tmp = processCommand( cmd );
+		if ( tmp instanceof Error ) {
+			debug( 'Unable to process command.' );
+			debug( 'Error: %s', tmp.message );
+			debug( 'Attempting to detect multi-line input...' );
+			if ( hasMultilineError( cmd, AOPTS ) ) {
+				debug( 'Detected multi-line input. Waiting for additional lines...' );
+				this._multiline.active = true;
+
+				// Insert a newline:
+				this._lineIndex += 1;
+				this._cmd.splice( this._lineIndex, 0, '' );
+				this._lines.splice( this._lineIndex, 0, '' );
+
+				// Display next prompt:
+				displayPrompt( this._repl, false );
+				return;
+			}
```

Comment:
@kgryte I was thinking, if we can refactor this.

Currently say we are at a line:

```
fo<|>o(
```

hitting CTRL+E:

```
fo
<|>o(
```

but hitting ENTER instead:

```
foo(
<|>
```

This is because we were deciding if the input is a possible multi-line candidate after the line event was triggered, meaning we were already on the next line, whereas with the keybinding we already know, it's a multiline trigger, so we accordingly handle buffers and cutting the line.

I propose, moving the multi-line candidate detection to a `beforeKeypress` handler, where we check before the ENTER keypress event, and so we know before the `line` event on whether to execute the expression or enter multi-line and can accordingly handle like we do with CTRL+E.

Reply:
Yes, that makes sense to me.